### PR TITLE
HPCC-12904 Introduce a heapHWM to optimize large allocations

### PR DESCRIPTION
**Pull request by ghalliday**
This significantly improves the perfromance of one of the unit tests
(testBitmapThreading) which simulates aggressively allocating huge memory
blocks.  It is unlikely to make a big difference to current queries,
but will be significant when multiple thor slaves are running in the
same process.

Signed-off-by: Gavin Halliday gavin.halliday@lexisnexis.com

### CODE REVIEW

**Comment by ghalliday** [comment]
@richardkchapman @mckellyln please review.

**Comment by AttilaVamos** [comment]
Automated Smoketest Build: success
HPCC Start: OK
HPCC Stop: OK

File: roxie/roxiemem/roxiemem.cpp [hunk 1]
```diff
@@ -748,6 +783,9 @@ static void clearBits(unsigned start, unsigned len)
 
 static void *subrealloc_aligned(void *ptr, unsigned pages, unsigned newPages)
 {
+    //If this function shrinks the size of the block, then the call to subfree_aligned() may increase heapHWM.
+    //If this function increases the size of the block, it is possible heapHWM could increase - it will be
```
Comment (richardkchapman) [inline roxie/roxiemem/roxiemem.cpp]: Do you mean 'heapHWM could decrease'?

Comment (ghalliday) [inline roxie/roxiemem/roxiemem.cpp]: Yes - updated

**Pull request merged**